graphql_compiler/fast_introspection.py

```diff
@@ -0,0 +1,300 @@
+        ),
+        "enumValues": (
+            [_get_enum_value(enum_value) for enum_value in enum_values] if enum_values else None
+        ),
+        "possibleTypes": (
+            [_get_type_ref(possible_type) for possible_type in possible_types]
+            if possible_types
+            else None
+        ),
+    }
+
+
+def _get_directive(directive: GraphQLDirective) -> Dict[str, Any]:
+    """Compute data for `directives` field of the introspection query for a particular directive."""
+    return {
+        "name": __Directive.fields["name"].resolve(directive, None),
+        "description": __Directive.fields["description"].resolve(directive, None),
+        "locations": [
+            __DirectiveLocation.serialize(location)
+            for location in __Directive.fields["locations"].resolve(directive, None)
+        ],
+        "args": [
+            _get_input_value(arg) for arg in __Directive.fields["args"].resolve(directive, None)
+        ],
+    }
+
+
+def _execute_fast_introspection_query(schema: GraphQLSchema) -> ExecutionResult:
+    """Compute the GraphQL introspection query."""
+    response_types = []
+    for type_ in __Schema.fields["types"].resolve(schema, None):
+        response_types.append(_get_full_type(type_, schema))
+
+    response_directives = []
+    for directive in __Schema.fields["directives"].resolve(schema, None):
+        response_directives.append(_get_directive(directive))
+
+    query_type = __Schema.fields["queryType"].resolve(schema, None)
+    mutation_type = __Schema.fields["mutationType"].resolve(schema, None)
+    subscription_type = __Schema.fields["subscriptionType"].resolve(schema, None)
+
+    response: Dict[str, Any] = {
+        "queryType": {"name": __Type.fields["name"].resolve(query_type, None)},
+        "mutationType": (
+            {"name": __Type.fields["name"].resolve(mutation_type, None)} if mutation_type else None
+        ),
+        "subscriptionType": (
+            {"name": __Type.fields["name"].resolve(subscription_type, None)}
+            if subscription_type
+            else None
+        ),
+        "types": response_types,
+        "directives": response_directives,
+    }
+    response_payload = {"__schema": response}
+
+    return ExecutionResult(data=response_payload, errors=None)
+
+
+def try_fast_introspection(schema: GraphQLSchema, query: str) -> Optional[ExecutionResult]:
+    """Compute the GraphQL introspection query if query can be computed fastly.
+
+    This can be used in replacement of `graphql_sync`. It automatically checks if the query
+    is the expected introspection query in this module; if the query is not the expected
+    introspection query, this function returns None, meaning the query cannot be computed fastly
+    with this module. Otherwise, it will execute the introspection query fastly and return
+    the computed result in the form of a GraphQL ExecutionResult.
```

Comment:
An (unfortunately poorly-documented) convention we do our best to stick to is that if a function requires a multi-line docstring, it should also always have explicit `Args / Returns` sections. For example: https://github.com/kensho-technologies/graphql-compiler/blob/main/graphql_compiler/compiler/compiler_frontend.py#L739